### PR TITLE
Move ValueNetworkMixin to torch_policy.py, make it switch implementat…

### PR DESCRIPTION
**Pull request by roireshef**
Move ValueNetworkMixin to torch_policy.py, make it switch implementations (baseline vs no baseline) based on use_critic, and add use_gae<->use_critic validation to all algos using ValueNetworkMixin

please see https://github.com/ray-project/ray/issues/15715 for description

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mvindiola1** [comment]
@roireshef and @sven1977,

The tf mixin also needs an update to "use_critic" match here:  
https://github.com/ray-project/ray/blob/dde7cbd2885765a9fc61b05daca4f6f9973aed10/rllib/agents/ppo/ppo_tf_policy.py#L276

File: rllib/policy/torch_policy.py [hunk 1]
```diff
@@ -933,6 +933,42 @@ def on_global_var_update(self, global_vars):
...
+            # input_dict.
+
+            def value(**input_dict):
+                input_dict = SampleBatch(input_dict)
+                input_dict = self._lazy_tensor_dict(input_dict)
+                model_out, _ = self.model(input_dict)
+                # [0] = remove the batch dim.
+                return self.model.value_function()[0].item()
+
+        # When not doing GAE, we do not require the value function's output.
```
Comment (mvindiola1) [inline rllib/policy/torch_policy.py:962]: This comment should be changed from "doing gae" to "using a critic"

**Comment by mvindiola1** [comment]
One last thing. I do not see a "use_critic" key in marwil's config. I think your validation  method for it will fail. Marwil does not actually use gae it only computes the cumulative return but it needed that key in order for the Torch ValueNetworkMixin to return a value estimate. I think you can safely change the name of that key from use_gae to use_critic.

**Comment by roireshef** [comment]
@mvindiola1 @sven1977 - hi guys, sorry for being so late here. I missed that it didn't go in. I've just went over the PR feedback and implemented the required changes. Thanks!

**Review by gjoliver — COMMENTED** [review]
I am actually a big fan of this PR.
I think it's really nice to have feature parity of ValueNetworkMixin for all the agents.

The lint error is a bit concerning though. If you are not making any actual logic changes, I guess we can merge the latest master again, trigger a new CI run, and see if the error is consistent.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by avnishn** [comment]
@gjoliver @roireshef should we still attempt to merge this pr?

**Comment by roireshef** [comment]
> @gjoliver @roireshef should we still attempt to merge this pr?

sure, why not.

**Comment by gjoliver** [comment]
Hi roireshef, thanks for the PR again.
I am gonna close this for now, since the team is deep down in the policy land to clean up a few things, and we will do this instead, so you don't have to bring this up to date, etc.
Thanks again.

**Pull request closed without merging**